tests/training/optimizer_test.py

```diff
@@ -77,6 +80,171 @@ def test_optimizer_parameter_groups(self):
         assert len(param_groups[2]["params"]) == 2
 
 
+class TestRegexOptimizer(AllenNlpTestCase):
+    def setup_method(self):
+        super().setup_method()
+        self.instances = SequenceTaggingDatasetReader().read(
+            self.FIXTURES_ROOT / "data" / "sequence_tagging.tsv"
+        )
+
+        vocab = Vocabulary.from_instances(self.instances)
+        self.model_params = Params(
+            {
+                "text_field_embedder": {
+                    "token_embedders": {"tokens": {"type": "embedding", "embedding_dim": 5}}
+                },
+                "encoder": {"type": "lstm", "input_size": 5, "hidden_size": 7, "num_layers": 2},
+            }
+        )
+        self.model = SimpleTagger.from_params(vocab=vocab, params=self.model_params)
+
+    def test_multiple_optimizers(self):
+        optimizer_params = Params(
+            {
+                "type": "regex",
+                "optimizers": {
+                    "embedder": {"type": "adam", "lr": 2},
+                    "encoder": {"type": "adam", "lr": 3},
+                    "default": {"type": "adam", "lr": 1},
+                },
+                "parameter_groups": [
+                    [
+                        ["^text_field_embedder"],
+                        {
+                            "optimizer_name": "embedder",
+                            "betas": (0.9, 0.98),
+                            "lr": 2,
+                            "weight_decay": 0.01,
+                        },
+                    ],
+                    [["^encoder.*bias"], {"optimizer_name": "encoder", "lr": 0.001}],
+                    [["^encoder.*weight"], {"optimizer_name": "encoder", "lr": 0.002}],
+                    [["^tag_projection_layer.*.weight$"], {"lr": 5}],
```

Comment:
I changed the format this expects a little bit. I didn't go quite as far as @epwalsh suggested, but far enough to make `Lazy` work. I think it's neat that the `parameter_groups` syntax stays the same compared to other optimizers (though I think that format is a little insane).

Reply:
> I wanted to use Lazy to create the inner optimizers. It's a problem to call into the `from_params` logic from a constructor, but it's also not obvious how to use `Lazy` to accomplish the same thing. Then things spiraled out of control a bit and I spent two days getting this to work the same way you had it already ...

That's cool that you got it to work with `Lazy`, I had initially used a `List` of `optimizers` as otherwise keying the optimizers on some name would throw an error that I hadn't registered that class if I recall correctly. By using `Lazy` it means you can construct the inner optimizers when you are ready which I think is a much better way of doing it. Thanks so much for updating all of the code, it's great to see it being done in a way I didn't think of! and sorry if I left you with a template that might have not been that flexible!

Reply:
Your template was great, saved me a ton of time. Especially the tests! I had to learn about `Lazy` first too.

It's way easier to fill in some function signatures than to come up with the whole structure yourself :-)